qcodes/dataset/dond/do_nd.py

```diff
@@ -34,104 +35,438 @@
-        ) = _extract_paramters_by_type_and_group(measurement_name, params_meas)
+        ) = self._extract_parameters_by_type_and_group(params_meas)
+        self.measurement_names = self._create_measurement_names(measurement_name)
+
+    def _create_measurement_names(
+        self, measurement_name: str | Sequence[str]
+    ) -> tuple[str, ...]:
+        if isinstance(measurement_name, str):
+            return (measurement_name,) * len(self._grouped_parameters)
+        else:
+            if len(measurement_name) != len(self._grouped_parameters):
+                raise ValueError(
+                    f"Got {len(measurement_name)} but should create {len(self._grouped_parameters)} datasets."
+                )
+            return tuple(measurement_name)
 
     @property
     def measured_all(self) -> tuple[ParamMeasT, ...]:
         return self._measured_all
 
     @property
-    def grouped_parameters(self) -> dict[str, ParameterGroup]:
+    def grouped_parameters(self) -> tuple[tuple[ParamMeasT, ...], ...]:
         return self._grouped_parameters
 
     @property
     def measured_parameters(self) -> tuple[ParameterBase, ...]:
         return self._measured_parameters
 
+    @staticmethod
+    def _extract_parameters_by_type_and_group(
+        params_meas: Sequence[ParamMeasT | Sequence[ParamMeasT]],
+    ) -> tuple[
+        tuple[ParamMeasT, ...],
+        tuple[tuple[ParamMeasT, ...], ...],
+        tuple[ParameterBase, ...],
+    ]:
+        measured_parameters: list[ParameterBase] = []
+        measured_all: list[ParamMeasT] = []
+        single_group: list[ParamMeasT] = []
+        multi_group: list[tuple[ParamMeasT, ...]] = []
+        grouped_parameters: tuple[tuple[ParamMeasT, ...], ...]
+        for param in params_meas:
+            if not isinstance(param, Sequence):
+                single_group.append(param)
+                measured_all.append(param)
+                if isinstance(param, ParameterBase):
+                    measured_parameters.append(param)
+            elif not isinstance(param, str):
+                multi_group.append(tuple(param))
+                for nested_param in param:
+                    measured_all.append(nested_param)
+                    if isinstance(nested_param, ParameterBase):
+                        measured_parameters.append(nested_param)
+        if single_group:
+            grouped_parameters = (tuple(single_group),)
+        if multi_group:
+            grouped_parameters = tuple(multi_group)
+        return tuple(measured_all), grouped_parameters, tuple(measured_parameters)
+
+
+# idealy we would want this to be frozen but then postinit cannot calulate all the parameters
+# https://stackoverflow.com/questions/53756788/
+@dataclass(frozen=False)
+class _SweapMeasGroup:
+    sweep_parameters: tuple[ParameterBase, ...]
+    measure_parameters: tuple[
+        ParamMeasT, ...
+    ]  # todo should this be all or only Parameters?
+    experiment: Experiment | None
```

Comment:
i didn't see where the experiment is used from this dataclass, so perhaps lets remove it? the experiment is only needed for passing to the measurement context, so since this class already gets a context then experiment is already there and hence should not be needed
```suggestion
```